flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala

```diff
@@ -358,20 +358,27 @@ abstract class TableEnvironment(val config: TableConfig) {
     * Registers an [[AggregateFunction]] under a unique name. Replaces already existing
     * user-defined functions under this name.
     */
-  private[flink] def registerAggregateFunctionInternal[T: TypeInformation, ACC](
+  private[flink] def registerAggregateFunctionInternal[T: TypeInformation, ACC: TypeInformation](
       name: String, function: AggregateFunction[T, ACC]): Unit = {
     // check if class not Scala object
     checkNotSingleton(function.getClass)
     // check if class could be instantiated
     checkForInstantiation(function.getClass)
 
-    val typeInfo: TypeInformation[_] = implicitly[TypeInformation[T]]
+    val resultTypeInfo: TypeInformation[_] = implicitly[TypeInformation[T]]
+    val accTypeInfo: TypeInformation[_] = implicitly[TypeInformation[ACC]]
 
     // register in Table API
     functionCatalog.registerFunction(name, function.getClass)
 
     // register in SQL API
-    val sqlFunctions = createAggregateSqlFunction(name, function, typeInfo, typeFactory)
+    val sqlFunctions = createAggregateSqlFunction(
```

Comment:
@fhueske thanks for your explanation. I tried to wrap the UDAGG in an `AggregateFunctionWrapper`, but find that it is not an easy way. Because I can't override the user-defined contract methods, such as `accumulate`, `retract`, `merge`. And in code generation, we generate the acc type class depend on `createAccumulator` method return type, but the return type of `createAccumulator` of `AggregateFunctionWrapper` can only be `Any` which will result in error.

In addition, I plan to support composite result type for UDAGG. This also need the return type not only to be known to Calcite for semantic validation but also to be known for code generation.